examples/file_upload/src/lib.rs

```diff
@@ -80,7 +80,7 @@ impl Component for Model {
                 </div>
                 <div>
                     <label>{ "By chunks" }</label>
-                    <input type="checkbox" checked=flag onclick=|_| Msg::ToggleByChunks />
+                    <input type="checkbox" checked=flag onclick=self.link.send_back(|_| Msg::ToggleByChunks) />
```

Comment:
While I think that `send_back` is fine, It might be worth considering just renaming it to `callback` or `make_callback` to make it painfully obvious what is happening now that this function is being used in many more places.

I think `onclick=self.link.callback(|_| Msg::ToggleByChunks)` is more readable and understandable to new users then the present naming scheme, which seems to be named more according to its implementation details rather than its purpose to users.

I propose the following renames - which might be better addressed in a separate PR, if at all:
* `send_back` -> `callback`
* `send_back_batch` -> `batch_callback`
* `send_self` -> `process` | `immediate` | `queue_message`
* `send_future` -> `future` (but this is being moved elsewhere anyways)

Reply:
If it's possible to access the `link` via the `Component` trait, it would be great to have something like `self.callback` or `self.message` provided by the trait as well (accessing the `ComponentLink` behind the scenes).

Reply:
Yeah, the ideal ergonomic solution would be:
```rust
onclick=self.callback(|_| Msg::ToggleByChunks)
```
And people would be free to create traits to add methods that make it even more terse if they so desired:
```rust
onclick=self.cb(|_| Msg::ToggleByChunks)
```

Reply:
In the event that Component links can't be gotten at arbitrary points, in order to add `self.callback()` as a short option , a `Linked` trait could be added to Yewtil, where users could define or derive which one of their fields is a ComponentLink, and then provide `callback`/`batch_callback` methods for the component instance.

Like:
```rust
#[derive(Linked)]
pub struct Model {
    #[link]
    link: ComponentLink<Self>
}
//...later
let cb = self.callback(|_| Msg::Something);

Reply:
Leveraged some macro magic in this draft PR: https://github.com/jstarry/yew/pull/3/files

Thinking it might be too magic though 🤔 let me know what y'all think!

Reply:
I'm impressed, but I don't have an affinity for using function-positioned macros to define structs, in that they don't end up immediately looking like struct declarations, and they tend to confuse IDE's syntax highlighting.

If I recall correctly, `Scope`s have a default instance created within a `html!` block, which is effectively attached to the "real" shared state later when the Html is integrated into the existing vdom state.
Could the same be done for `ComponentLink`s?
Being able to create links from thin air would be a nice shortcut around this problem.

I think either that, providing links on lifecycle hooks (or maybe just `view`?), or attribute-based or automatic deduction of the link field within the above proposed `Linked` derive trait, might be acceptable alternatives.

Reply:
I think `let callback = self.link.callback;` or `let cb = self.link.callback;` becoming an idiom solves the problem of verbosity within `html!{}` blocks. I don't know if something short of your proposal will make removing boilerplate from creation possible though.

Maybe extending `Linked` could allow implementing a `LinkedComponent` trait that doesn't have a `create` method. `Properties` would have to be annotated, link could be automatically be deduced based on its type. This would all live in yewtil.

eg:

```rust
#[derive(Linked)]
pub struct Model {
    link: ComponentLink<Self>,
    #[properties] 
    props: Props,
    some_value: usize
}

impl LinkedComponent {
    type Message = Msg;
    fn update(&mut self, msg: Self::Message) -> ShouldRender {}
// etc...
}

// ---
// The auto-derived create would look like:
fn create(props: Props, link: ComponentLink<Self>) -> Self {
    Self {
        link: link,
        props: props,
        some_value: Default::default()
    }
}
```

Reply:
After working with this a bit more, the boilerplate bits aren't actually too too bad. It feels repetitive when you're creating a new component, but the frequency of doing that is low, and it's easily circumvented with a bit of text-editor config for autocompletes and the like. 

The one thing that I could see being useful (although I'm unsure how feasible it is) could be if the `html!` macro accepted a `link` (callback?) to bind event callbacks as was currently done. E.g.:

```rust
fn view(&self) -> Html {
    html!({
        <button onclick=|_| Msg::Clicked>{"Click me!"}</button>
    }, self.link)
}
```

This would have to be an optional thing however (for component structs that don't have an explicitly stored `ComponentLink` member) and it wouldn't work if there was a mix of `ComponentLinks` that should be used in the view such as if making a table component where the table row can be selected (handled by the `Table` component), but the _contents_ of the table row could contain a button that when clicked, needs to message the top level component (I can't remember whether we're calling that the base or parent... the most outer one though). 

Might be best to just allow for some yew-idioms to develop and see what patterns emerge before making more drastic changes though.

Reply:
I love the idea of providing a link to the `html!` macro - as long as falling back to `onclick=self.link.callback(etc)` is allowed if it isn't provided. The only minor issue I have with it is that it isn't possible to propagate that through inner control structures (I think), or html blocks defined elsewhere:
eg:
```rust
html! ({
    <>
    <div> {"text"} </div>
    {  
        if cond {
            html!({...}, self.link)
        } else {
            html!({...}, self.link)
        }
    }
}, self.link) // doesn't propagate to inner html!s
``` 

It would undoubtedly save a bunch of people's code from breaking because it would allow them to keep the same old html syntax by just adding a `self.link` to the end of their html, but at the same time, for branchy code with sparse callbacks, it might get in the way more than it would help, as well as introducing multiple ways to accomplish the same thing - leading to confusion. There is also the question about whether adding the additional complexity to the parser is worth it.

-------
I am in agreement that link being required everywhere isn't as much of a problem as I initially made it out to be.

Reply:
How about we follow up on the ergonomics in another issue?